### ISSUE TITLE
defold.Buffer or defold.types.Buffer

### ISSUE THREAD

**Issue by AsakuraMizu** [hxdefold/hxdefold#29]
`defold.Buffer.create` returns a `defold.Buffer` object. Other methods also require a `defold.Buffer` object as argument.

But there's a `defold.types.Buffer` class.When I write `import defold.types.*` and `import defold.Buffer`, `Buffer` does not behavior as excepted.

**Comment by haath** [comment]
As far as I can tell the difference is that `defold.Buffer` is a static class, which exists to forward calls to `buffer.copy_buffer()` etc.
While `defold.types.Buffer` is an *instance* class meant to be returned by other methods, and then **used as an argument** for calls to such methods like `copy_buffer()`.

So we do indeed have a problem here, because the `buffer` methods need `defold.Buffer` as an argument, leaving `defold.types.Buffer` completely useless...

The *correct* thing would be to switch all the parameter types to `defold.types.Buffer`, but then we're left with the name conflict.

The naming conflict really comes from just following the Defold API naming, so I think we should just remove `defold.types.Buffer`. Right @nadako ?

**Comment by AsakuraMizu** [comment]
> The _correct_ thing would be to switch all the parameter types to `defold.types.Buffer`, but then we're left with the name conflict.

Yes you are right, there is a **name conflict**. But instead of just removing `defold.types.Buffer`, I suggest using some tricks to avoid name conflict e.g. renaming `defold.types.Buffer` to `defold.types.TBuffer` or `defold.types.BufferType`.

**Comment by haath** [comment]
We do prefer to stick with whatever the Defold documentation calls things so far for clarity, since the library is mostly just a layer of externs. Removing it would be cleaner, but it could be wrong semantically since `defold.Buffer` would then be a mixed static/instance class which wouldn't match 100% with the Defold API.

Will think about it some more, maybe renaming is better.

---

In your own project you're probably better off creating a wrapper class anyway.

For example:

```haxe
import defold.Buffer;
typedef BufferType = defold.types.Buffer;

class MyBuffer
{
    var buf: BufferType;

    public function new(elemCount: Int, declaration: lua.Table<Int,BufferStreamDeclaration>)
    {
        buf = Buffer.create(elemCount, declaration);
    }

    public inline function getStream(name: HashOrString): BufferStream
    {
        return Buffer.get_stream(buf, name);
    }

    // etc...
}
```

Which we could probably also include here at some point, there are some open issues from nadako about switching to a more OOP approach instead of just having externs but I haven't gotten started on anything yet.

**Comment by haath** [comment]
The main issue here was fixed by #31 which correctly defined the usage of `defold.types.Buffer` in the function arguments.

As long as `hxdefold` provides only externs, I think this is a good solution and allows the user to use the complete API.
So I'll close this issue for now.
